### PR TITLE
Issue996 hydronictest

### PR DESCRIPTION
**Pull request by Mathadon**
this is for #996 

further revisions are still required

@dhblum can you comment further on the remaining features that still have to be added?

### CODE REVIEW

**Comment by dhblum** [comment]
I emailed to you Review 2 using BOPTEST review guide.

Discussion of any additional remaining features forthcoming soon.

**Comment by damienpicard** [comment]
So, shall I still wait before looking at it then?

**Comment by Mathadon** [comment]
@damienpicard indeed no need to merge/review this yet

**Comment by JavierArroyoBastida** [comment]
@dhblum, @Mathadon the BOPTEST test case associated to this model is about to be ready. I've basically gone through [these points](https://github.com/ibpsa/project1-boptest/pull/201#issuecomment-625146507) and now the FMU contains the weather file thanks to [this pull request](https://github.com/ibpsa/modelica-ibpsa/pull/1374). Only thing I'm not yet convinced of is the baseline controller which is very naive right now: it's just hysteresis switching on/off the pump of the emission system, but the supply water temperature from the boiler is fixed, which leads to large temperature overshoots. If you agree I'll add modulation of boiler supply water temperature to the baseline controller to get a smother and more realistic behavior.

**Comment by dhblum** [comment]
Previously, @Mathadon mentioned this: https://github.com/open-ideas/IDEAS/pull/1111#discussion_r386683185, that the thermostat switches the system on/off to meet setpoint.  In that thread, we talked about adding the ability to overwrite supply water temperature as well.  I understand there could be a heating curve for the boiler.  Is it often the case that a RBC/thermostat would change the supply water temperature with a heating curve in a simple system like this?

**Comment by JavierArroyoBastida** [comment]
I think a heating curve based on outdoor temperature is not that common to have for such a simple system. Controllers in simple houses are rather based on temperature differences between the set point and the actual temperature, but still provide a smoother behavior than the one showed in the current implementation. My intuition is that these controllers not only use hysteresis but also some other rules to improve performance, am I right @Mathadon? We can leave the baseline controller like this, but just want to ensure that we provide a realistic control along with the model.

**Comment by Mathadon** [comment]
> Previously, @Mathadon mentioned this: [#1111 (comment)](https://github.com/open-ideas/IDEAS/pull/1111#discussion_r386683185), that the thermostat switches the system on/off to meet setpoint. In that thread, we talked about adding the ability to overwrite supply water temperature as well. I understand there could be a heating curve for the boiler. Is it often the case that a RBC/thermostat would change the supply water temperature with a heating curve in a simple system like this?

my 100 EUR thermostat does it at home, so I do think that this is happening a lot nowadays.

**Comment by Mathadon** [comment]
> I think a heating curve based on outdoor temperature is not that common to have for such a simple system. Controllers in simple houses are rather based on temperature differences between the set point and the actual temperature, but still provide a smoother behavior than the one showed in the current implementation. My intuition is that these controllers not only use hysteresis but also some other rules to improve performance, am I right @Mathadon? We can leave the baseline controller like this, but just want to ensure that we provide a realistic control along with the model.

My thermostat is not just using a heating curve. When the schedule switches to a higher temperature, the thermostat first uses full power/temperature. The temperature only decreases when approaching the set point. Looks like a PI type of control could be more appropriate.

**Comment by JavierArroyoBastida** [comment]
I'm proposing [here](https://github.com/open-ideas/IDEAS/pull/1146)  a more complete control scheme that uses a PI for modulating the boiler supply water temperature on top of the pump hysteresis. Below you can see the comparison of both controllers: the pink line is the indoor temperature when using the one with only hysteresis and green line is the indoor temperature when using hysteresis+PI. The bottom graph shows suplly water temperature of the boiler for the second case between 20 and 80 degrees Celsius. (first case the supply water temperature is constant at 60 degrees Celsius). I think it's more realistic. Let me know what you think.

SingleZoneResidentialHydronic

**Comment by dhblum** [comment]
Thanks @JavierArroyoBastida.  Does the system have enough capacity in the new case?  The setpoint isn't reach for many of the days.

**Comment by JavierArroyoBastida** [comment]
@dhblum I didn't modify the heat capacity, but now the temperature is always at the lowest allowed bound and the thermal inertia and large setback prevent the system to reach the setpoint when switching between unoccupied to occupied periods. It can be resolved by either increasing thermal capacity or internally adding an offset to the baseline controller that does not allow such low temperatures during unoccupied periods.

**Comment by dhblum** [comment]
I'm not sure what you mean by "internally adding an offset to the baseline controller that does not allow such low temperatures during unoccupied periods."  By thermal capacity, do you mean heating capacity of the system, or heat capacitance (e.g. thicker concrete floor) to the zone?  I think I would prefer increasing heating capacity of system of the options.

**Comment by Mathadon** [comment]
The PI implementation seems to work well! Nice.

The set points not being reached is physical. I'd increase the setback level to 16C, which is what I use at home. The system is probably sized quite small if it takes so long to reach the set point, so I'd increase the normal powers by ~30%?

**Comment by JavierArroyoBastida** [comment]
Right now there is a small offset of 0.2C that is added to the heating setpoint preventing from discomfort due to slight oscilations around the PI controller temperature reference. So the PI reference is not the heating setpoint directly but the heating setpoint + 0.2C. By "_adding an offset to the basline controller that does not allow such low temperatures during unoccupied periods_" I meant to increase the offset during unoccupied periods, e.g. to 1C or 2C, not to prevent from discomfort due to small temperature oscilations, but to compensate for the large setback being used that clearly causes discomfort during occupancy switches. 

Another options is to directly increase the setback level as @Mathadon proposes, but I am more in favour of increasing the offset during unoccupied periods since the controller settings should not affect how discomfort is defined in the test case: if there is a controller that sees value on lowering down temperatures to 14C we shouldn't deprive it because the baselince controller does not work well with this setback. In any case, I agree that increasing normal powers by ~30% may be a good idea (by thermal capacity I was meaning heating capacity, but actually none of heat capacity nor thermal capacitance have been changed so far).

**Comment by Mathadon** [comment]
>  if there is a controller that sees value on lowering down temperatures to 14C we shouldn't deprive it because the baselince controller does not work well with this setback

that's a good point. So you want to change the PI set point but not the actual limit? Ok for me!

**Comment by JavierArroyoBastida** [comment]
Yes exactly. Ok I'll wait for @dhblum's opinion and then come up with a proposal.

**Comment by dhblum** [comment]
Just double checking, the comfort limit in unoccupied times for the test case is set to 15 C I think from the plots, not 14 C as you mention?  

I guess this is where something like an 'optimal start' algorithm comes in.  Being able to heat/cool a building adequately for when people arrive.  But @Mathadon, when you say at home you set the setpoint to 16 C, that's because that is your comfort limit?  Or are you doing that so that when you increase it during normal hours, the system can reach the normal setpoint faster (or even at all in the case of many days in the plots)?  I ask the difference, because by keeping the test case min at 15 C, but raising the setback on the PI to 16 C just so it can reach the normal setpoint at 21 C would be like the latter purpose, which I don't think people do.  Rather, I think people set the setback to the minimum comfortable temperature.  That would indicate to me actually raising the test case setback temperature to 16 C.  Or, again, just increase the capacity of the system if 15 C is really what we want the minimum temperature to be for the test case.

**Comment by JavierArroyoBastida** [comment]
Indeed, comfort limit is set to 15 C, not 14 C, that was a mistake. 
Any option between raising setback temperature, PI reference, or heating capacity is fine by me. 
If comfort is defined at 16 C during unoccupied periods then that's the setback we should use. If after that the setpoint is still not reached then any of raising PI reference or heating capacity should be used, but not increasing the seback anymore. What I mean is that comfort definition should come before control logic.

**Comment by dhblum** [comment]
I think PI reference should equal setback temperature, and system sized appropriately to meet setpoints.  Maybe not immediately, but within a reasonable time.

**Comment by dhblum** [comment]
Actually, let me rephrase a bit.  I agree with you for the most part.  I think I'm just making a stronger preference for increasing the system size to meet setpoints, rather than assume that the system is undersized for the baseline controller that would likely come with such a system.  To me, this assumes some design flaw and arbitrarily puts the baseline controller at a disadvantage.

**Comment by dhblum** [comment]
So define comfort at 16 C, PI reference at 16 C (+ whatever deadband), increase system size by ~30%?

**Comment by JavierArroyoBastida** [comment]
Increasing nominal heating capacity of boiler and radiators from 3 kW to 5 kW seems to work just fine. I've also decreased proportional constant and deadband from 0.2 degC to 0.1 degC as seems to improve the baseline. I've updated https://github.com/open-ideas/IDEAS/pull/1146 and the testcase documentation with these changes. These are the simulation results:

SingleZoneResidentialHydronic

If you agree with the changes and have no further comments I'll update https://github.com/ibpsa/project1-boptest/pull/201 by re-compiling the model and updating references. After that the testcase should be ready to merge.

**Comment by dhblum** [comment]
The new results look good to me!  I don't have further comments on them.  Thanks @JavierArroyoBastida!

**Comment by Mathadon** [comment]
Looks good indeed :) Is this ready for a review before merging?

**Comment by JavierArroyoBastida** [comment]
Great! I'd say the model is ready for review but the BOPTEST case still needs an update. I'll let you know once that's done for a complete review.

**Comment by Mathadon** [comment]
This is ready to be merged once the unit tests pass.
